addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/http/ComparableResponse.java

```diff
@@ -0,0 +1,521 @@
+        }
+        return 1f;
+    }
+
+    /**
+     * Return a heuristic evaluation between the number of words of the 2 responses. 0 means very
+     * different and 1 very similar.
+     *
+     * @param response1 The first response to compare
+     * @param response2 The second response to compare
+     */
+    public static float wordCountHeuristic(
+            ComparableResponse response1, ComparableResponse response2) {
+        float nWords1 = response1.body.split("\\s+").length;
+        float nWords2 = response2.body.split("\\s+").length;
+
+        if (nWords1 > nWords2) {
+            return nWords2 / nWords1;
+        } else if (nWords1 < nWords2) {
+            return nWords1 / nWords2;
+        }
+        return 1f;
+    }
+
+    /**
+     * Return a heuristic evaluation between the number of lines of the 2 responses. 0 means very
+     * different and 1 very similar.
+     *
+     * @param response1 The first response to compare
+     * @param response2 The second response to compare
+     */
+    public static float lineCountHeuristic(
+            ComparableResponse response1, ComparableResponse response2) {
+        float nLinesR1 = response1.body.split("\r\n|\r|\n").length;
+        float nLinesR2 = response2.body.split("\r\n|\r|\n").length;
+
+        if (nLinesR1 > nLinesR2) {
+            return nLinesR2 / nLinesR1;
+        } else if (nLinesR1 < nLinesR2) {
+            return nLinesR1 / nLinesR2;
+        }
+        return 1f;
+    }
+
+    /**
+     * Return a heuristic evaluation between the 2 Responses headers. 0 means very different and 1
+     * very similar.
+     *
+     * @param response1 The first response to compare
+     * @param response2 The second response to compare
+     */
+    public static float headersCompareHeuristic(
+            ComparableResponse response1, ComparableResponse response2) {
+        if (response1.headers.isEmpty() && response2.headers.isEmpty()) {
+            return 1; // Both empty, equal
+        } else if (response1.headers.isEmpty() || response2.headers.isEmpty()) {
+            return 0; // One empty, very different
+        }
+        float equalHeadersContent = 0;
+        int r1HeaderCount = response1.headers.keySet().size();
+        int r2HeaderCount = response2.headers.keySet().size();
+        float numberHeaders = r1HeaderCount > r2HeaderCount ? r1HeaderCount : r2HeaderCount;
+
+        for (String key : response1.headers.keySet()) {
+            if (response2.getHeaders().containsKey(key)
+                    && (response2.getHeaders().get(key).equals(response1.headers.get(key))
+                            || DYNAMIC_HEADERS.contains(key))) {
+                equalHeadersContent += 1;
+            }
+        }
+        // Similar to (equalHeaders / numberHeaders * equalHeadersContent / equalHeaders)* 0.25 +
+        // 0.75;
+        return equalHeadersContent / numberHeaders;
+    }
```

Comment:
I started to change this because as soon as one header matched then the return was 1.0, however say we have a message with 1 header and another with 3, one of which match. Shouldn't that be 0.33 not 1.0?

This would be applicable to a number of the comparison methods. However, then the question becomes do we always make the 'first' thing the larger item/set? If that's not always the same message will it matter at all? 
(I think the comparison score will still be right if we "favor" the larger in each part but we should discuss/debate :wrinning: )

Reply:
@psiinon @thc202 @ricekot thoughts/feedback?

Reply:
@DiogoMRSilva maybe you have some thoughts on this?

Reply:
Hi @kingthorin happy to see this finally used on the official ZAP :)
> I started to change this because as soon as one header matched then the return was 1.0. However, say we have a message with 1 header and another with 3, one of which match. Shouldn't that be 0.33 not 1.0?

I agree that we need to use the number of headers of the request that has more headers.

> then the question becomes do we always make the 'first' thing the larger item/set? If that's not always the same message will it matter at all?

I think we should always use the Comparable message as the same element in the "operation" because we want to do the same comparison when comparing the X message to  more than 1 msg.

Reply:
> I think we should always use the Comparable message as the same element in the "operation" because we want to do the same comparison when comparing the X message to more than 1 msg.

What I meant here is that for each of the methods, not just the headers, should we be considering the larger "set" vs the smaller? (# of lines, # of words, etc)

Reply:
> What I meant here is that for each of the methods, not just the headers, should we be considering the larger "set" vs the smaller? (# of lines, # of words, etc)

yes I think so. Other solution would be to join the headers(or other parameter) names from both requests and check which percentage of them exist in both. Since these are heuristics I don't think there will be a 100% correct solution